src/Processors/Formats/Impl/ParquetBlockInputFormat.h

```diff
@@ -306,6 +314,17 @@ class ParquetSchemaReader : public ISchemaReader
     std::shared_ptr<parquet::FileMetaData> metadata;
 };
 
+class ParquetFileMetaDataCache : public CacheBase<String, parquet::FileMetaData>
+{
+public:
+    static ParquetFileMetaDataCache *  instance(UInt64 max_cache_entries);
+    void clear() {}
+
+private:
+    ParquetFileMetaDataCache(UInt64 max_cache_entries);
+    static std::mutex mutex;
```

Comment:
It appears to me `CacheBase` already has a mutex https://github.com/ClickHouse/ClickHouse/blob/94709d7a5a1ede65a9eed92ff9cbf73e28e62561/src/Common/CacheBase.h#L245.

Why do we need another one? And why is it static?

I think I am missing something. `ParquetFileMetaDataCache` is a static instance, shared among all instances. Why do you need to make this mutex static as well?

Reply:
Yes, basic double checked locking pattern of singleton initialization. 

I will replace this with the modern equivalent that uses `std::once_only.`

Reply:
Check my comment in https://github.com/Altinity/ClickHouse/pull/541/files#r1875950244. I might be missing something, tho

Reply:
Done.